engines/twp/twp.h

```diff
@@ -0,0 +1,247 @@
+	bool execSentence(Common::SharedPtr<Object> actor, VerbId verbId, Common::SharedPtr<Object> noun1, Common::SharedPtr<Object> noun2 = nullptr);
+
+	float getRandom() const;
+	float getRandom(float min, float max) const;
+
+private:
+	void update(float elapsedMs);
+	void draw(RenderTexture *texture = nullptr);
+	void exitRoom(Common::SharedPtr<Room> nextRoom);
+	void cancelSentence(Common::SharedPtr<Object> actor = nullptr);
+	void clickedAt(Math::Vector2d scrPos);
+	bool clickedAtHandled(Math::Vector2d roomPos);
+	void setShaderEffect(RoomEffect effect);
+	bool selectable(Common::SharedPtr<Object> actor);
+	void resetVerb();
+	Common::String cursorText();
+	Verb verb();
+	bool preWalk(Common::SharedPtr<Object> actor, VerbId verbId, Common::SharedPtr<Object> noun1, Common::SharedPtr<Object> noun2);
+	void updateTriggers();
+	void callTrigger(Common::SharedPtr<Object> obj, HSQOBJECT trigger);
+	Common::Array<ActorSwitcherSlot> actorSwitcherSlots();
+	ActorSwitcherSlot actorSwitcherSlot(ActorSlot *slot);
+	Scaling *getScaling(const Common::String &name);
+	void skipCutscene();
+
+private:
+	Vm _vm;
+
+public:
+	Graphics::Screen *_screen = nullptr;
+	GGPackSet _pack;
+	ResManager _resManager;
+	Common::Array<Common::SharedPtr<Room> > _rooms;
+	Common::Array<Common::SharedPtr<Object> > _actors;
+	Common::Array<Common::SharedPtr<ThreadBase> > _threads;
+	Common::Array<Common::SharedPtr<Task> > _tasks;
+	Common::Array<Common::SharedPtr<Callback> > _callbacks;
+	Common::SharedPtr<Object> _actor;
+	Common::SharedPtr<Object> _followActor;
+	Common::SharedPtr<Room> _room;
+	float _time = 0.f;
+	Common::SharedPtr<Object> _noun1;
+	Common::SharedPtr<Object> _noun2;
+	UseFlag _useFlag;
+	HSQOBJECT _defaultObj;
+	bool _walkFastState = false;
+	bool _holdToMove = false;
+	float _nextHoldToMoveTime = 0.f;
+	int _frameCounter = 0;
+	Common::SharedPtr<Lighting> _lighting;
+	Common::SharedPtr<Cutscene> _cutscene;
+	Scene _scene;
+	Scene _screenScene;
+	NoOverrideNode _noOverride;
+	InputState _inputState;
+	Camera _camera;
+	TextDb _textDb;
+	Dialog _dialog;
+	struct Cursor {
+		Math::Vector2d pos;
+		bool oldLeftDown = false;
+		bool leftDown = false;
+		int leftDownTime = 0;
+		bool oldRightDown = false;
+		bool rightDown = false;
+		int doubleClick = false;
+
+		void update() {
+			oldLeftDown = leftDown;
+			oldRightDown = rightDown;
+		}
+		bool isLeftDown() { return !oldLeftDown && leftDown; }
+		bool isRightDown() { return !oldRightDown && rightDown; }
+	} _cursor;
+	Hud _hud;
+	Inventory _uiInv;
+	ActorSwitcher _actorSwitcher;
+	AudioSystem _audio;
+	SaveGameManager _saveGameManager;
+	ShaderParams _shaderParams;
+	HotspotMarkerNode _hotspotMarker;
+	unique_ptr<FadeShader> _fadeShader;
+	LightingNode _lightingNode;
+
+private:
+	Gfx _gfx;
+	Preferences _prefs;
+	SentenceNode _sentence;
+	WalkboxNode _walkboxNode;
+	PathNode _pathNode;
+	Shader _bwShader;
+	Shader _ghostShader;
+	Shader _sepiaShader;
+};
+
+extern TwpEngine *g_engine;
```

Comment:
If I remember correctly I used `create_project` to generate the skeleton of the egine, and it was in this generated code.

Reply:
Right, I didn't remember this being part of the generated code. Technically it should be completely fine since the object is the exact same, you're just redefining it as the specific engine subclass, and only within that engine's namespace. When I was submitting an engine I got advised the same thing and I do find that when reading the code it communicates intent slightly better. I'm leaving the decision up to you, though.